src/Databases/DatabaseOnDisk.cpp

```diff
@@ -213,8 +218,47 @@ void DatabaseOnDisk::commitCreateTable(const ASTCreateQuery & query, const Stora
         Poco::File(table_metadata_tmp_path).remove();
         throw;
     }
+
+    try
+    {
+        /// If the table was detached permanently we will have a file with
+        /// .sql_detached suffix, which is not needed anymore since we attached the table back
+        auto table_metadata_file_detached = Poco::File(table_metadata_path + detached_suffix);
+        if (table_metadata_file_detached.exists())
+            table_metadata_file_detached.remove();
+    }
+    catch (...)
+    {
+        // It's not a big issue if we can't remove the .sql_detached file.
+        LOG_WARNING(log, getCurrentExceptionMessage(__PRETTY_FUNCTION__));
+    }
+}
+
+void DatabaseOnDisk::detachTablePermanently(const String & table_name)
+{
+    StoragePtr table = detachTable(table_name);
+
+    /// This is possible for Lazy database.
+    if (!table)
```

Comment:
But `table` is not actually used below, maybe we can just remove this `if`, so `DETACH PERMANENTLY` will work for `Lazy` db?

Reply:
ok. removed, will add some simple test.